### ISSUE TITLE
Forced indexHint raises error although mentioned index is used when hint is removed 

### ISSUE THREAD

**Issue by matcho** [arangodb/arangodb#20216]
## My Environment

* __ArangoDB Version__:        3.11.4 (also tested with 3.8.7)
* __Deployment Mode__:       Single
* __Deployment Strategy__:   systemd
* __Infrastructure__:               own
* __Operating System__:        Ubuntu 20.04
* __Total RAM in your machine__:        256G
* __Disks in use__:        SSD
* __Used Package__:              Ubuntu .deb

## Component, Query & Data

__Affected feature__:
AQL query

__AQL query (if applicable)__:
```
for tag in [ "a", "b", "c" ]
    for tc in tags_catalog OPTIONS { indexHint: "idx_1722945121587036160", forceIndexHint: true }
        filter tc.name == tag.name
        filter tc.public_display == true // remove/comment this → it works
        return tc
```

__AQL explain and/or profile (if applicable)__:
After removing `filter tc.public_display == true`
```
Query String (258 chars, cacheable: true):
 for tag in [ "a", "b", "c" ]
     for tc in tags_catalog OPTIONS { indexHint: "idx_1722945121587036160", forceIndexHint: true }
         filter tc.name == tag.name
         // filter tc.public_display == true // remove/comment this → it works
         return tc
 

Execution plan:
 Id   NodeType            Est.   Comment
  1   SingletonNode          1   * ROOT
  2   CalculationNode        1     - LET #2 = [ "a", "b", "c" ]   /* json expression */   /* const assignment */
  3   EnumerateListNode      3     - FOR tag IN #2   /* list iteration */
  8   IndexNode              3       - FOR tc IN tags_catalog   /* persistent index scan */    
  7   ReturnNode             3         - RETURN tc

Indexes used:
 By   Name                      Type         Collection     Unique   Sparse   Selectivity   Fields       Ranges
  8   idx_1722945121587036160   persistent   tags_catalog   true     false       100.00 %   [ `name` ]   (tc.`name` == tag.`name`)

Optimization rules applied:
 Id   RuleName
  1   use-indexes
  2   remove-filter-covered-by-index
  3   remove-unnecessary-calculations-2

Optimization rules with highest execution times:
 RuleName                                    Duration [s]
 use-indexes                                      0.00007
 interchange-adjacent-enumerations                0.00004
 remove-filter-covered-by-index                   0.00002
 reduce-extraction-to-projection                  0.00002
 remove-unnecessary-calculations-2                0.00001

72 rule(s) executed, 2 plan(s) created
```

Keeping `filter tc.public_display == true` but removing `OPTIONS { indexHint: "idx_1722945121587036160", forceIndexHint: true }`
```
Query String (258 chars, cacheable: true):
 for tag in [ "a", "b", "c" ]
     for tc in tags_catalog // OPTIONS { indexHint: "idx_1722945121587036160", forceIndexHint: true }
         filter tc.name == tag.name
         filter tc.public_display == true // remove/comment this → it works
         return tc
 

Execution plan:
 Id   NodeType            Est.   Comment
  1   SingletonNode          1   * ROOT
  2   CalculationNode        1     - LET #2 = [ "a", "b", "c" ]   /* json expression */   /* const assignment */
  3   EnumerateListNode      3     - FOR tag IN #2   /* list iteration */
 10   IndexNode              3       - FOR tc IN tags_catalog   /* persistent index scan */    FILTER (tc.`public_display` == true)   /* early pruning */   
  9   ReturnNode             3         - RETURN tc

Indexes used:
 By   Name                      Type         Collection     Unique   Sparse   Selectivity   Fields       Ranges
 10   idx_1722945121587036160   persistent   tags_catalog   true     false       100.00 %   [ `name` ]   (tc.`name` == tag.`name`)

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   move-calculations-up-2
  4   move-filters-up-2
  5   use-indexes
  6   remove-filter-covered-by-index
  7   remove-unnecessary-calculations-2
  8   move-filters-into-enumerate

Optimization rules with highest execution times:
 RuleName                                    Duration [s]
 use-indexes                                      0.00009
 interchange-adjacent-enumerations                0.00004
 remove-filter-covered-by-index                   0.00002
 move-calculations-up-2                           0.00002
 reduce-extraction-to-projection                  0.00002

72 rule(s) executed, 2 plan(s) created
```

__Dataset__:
Collection `tags_catalog` has 12 docs and the following index:
`20791424324	persistent	true	false	deduplicate: false, estimates: true	100.00%	name	idx_1722945121587036160`

## Steps to reproduce

1.  Execute query above: an error is raised `Query: AQL: could not use index hint to serve query; {"indexHint":{"forced":true,"type":"simple","hint":["idx_1722945121587036160"]}} (while optimizing plan)`
2.  Remove the hint or the filter clause that is not related to the index: query succeeds and explain shows that the index used is the one we required in the hint

__Problem__:
`indexHint` should be accepted

__Expected result__:
No error, query runs fine

Thank you,
Mathias

**Comment by jsteemann** [comment]
Interesting, it turned out that this issue has the same root cause as https://github.com/arangodb/arangodb/issues/20168:
the optimizer generates multiple plan permutations, and some of them are not able to use the index.

For example, here are the two plans generated by the optimizer for the query (can be made visible by passing the `allPlans: true` option when running explain):

```
Plan #1 of 2 (estimated cost: 3.00)

 Execution plan:
  Id   NodeType                  Est.   Comment
   1   SingletonNode                1   * ROOT
   2   CalculationNode              1     - LET #2 = [ "a", "b", "c" ]   /* json expression */   /* const assignment */
   4   EnumerateCollectionNode      0     - FOR tc IN tags_catalog   /* full collection scan  */   FILTER (tc.`public_display` == true)   /* early pruning */
   3   EnumerateListNode            0       - FOR tag IN #2   /* list iteration */
   5   CalculationNode              0         - LET #4 = (tc.`name` == tag.`name`)   /* simple expression */   /* collections used: tc : tags_catalog */
   6   FilterNode                   0         - FILTER #4
   9   ReturnNode                   0         - RETURN tc

 Indexes used:
  none

 Optimization rules applied:
  Id   RuleName
   1   move-calculations-up
   2   move-filters-up
   3   interchange-adjacent-enumerations
   4   move-calculations-up-2
   5   move-filters-up-2
   6   move-filters-into-enumerate

 78 rule(s) executed, 2 plan(s) created, peak mem [b]: 0, exec time [s]: 0.00048


Plan #2 of 2 (estimated cost: 5.00)

 Execution plan:
  Id   NodeType            Est.   Comment
   1   SingletonNode          1   * ROOT
   2   CalculationNode        1     - LET #2 = [ "a", "b", "c" ]   /* json expression */   /* const assignment */
   3   EnumerateListNode      3     - FOR tag IN #2   /* list iteration */
  10   IndexNode              0       - FOR tc IN tags_catalog   /* persistent index scan, index scan + document lookup */    FILTER (tc.`public_display` == true)   /* early pruning */   
   9   ReturnNode             0         - RETURN tc

 Indexes used:
  By   Name                      Type         Collection     Unique   Sparse   Cache   Selectivity   Fields       Stored values   Ranges
  10   idx_1722945121587036160   persistent   tags_catalog   false    false    false      100.00 %   [ `name` ]   [  ]            (tc.`name` == tag.`name`)

 Optimization rules applied:
  Id   RuleName
   1   move-calculations-up
   2   move-filters-up
   3   move-calculations-up-2
   4   move-filters-up-2
   5   use-indexes
   6   remove-filter-covered-by-index
   7   remove-unnecessary-calculations-2
   8   move-filters-into-enumerate

 78 rule(s) executed, 2 plan(s) created, peak mem [b]: 0, exec time [s]: 0.00048
```
So we have 2 plans, and the first one is not using an index scan but a full collection scan. Only the second plan uses an index at all.
The first plan is assigned a lower cost than the second plan, that's why the optimizer picks the first plan.
And as it does not use an index as dictated by the index hint, the query fails.

In my tests, the optimizer prefers the full scan over the index scan in case there are only few documents and/or very few different index entries. That means the same query may actually work if the data in the underyling collection/index changes.

It seems right now that the fix in https://github.com/arangodb/arangodb/pull/20201 also fixes this particular issue, but the fix has not been tested extensively and is thus still an open PR.

**Comment by jsteemann** [comment]
The issue will be fixed by the changes made in https://github.com/arangodb/arangodb/pull/20201.
That is a PR for our devel though, but once it is landed there we can backport the change to 3.11.

**Comment by jsteemann** [comment]
Will be fixed in 3.11 via https://github.com/arangodb/arangodb/pull/20249
